src/mem/arena.rs

```diff
@@ -337,14 +361,135 @@ unsafe impl<'arena> Arena for BumpArena<'arena> {
     }
 }
 
+/// A type that can serve as a buffer an arena can allocate from.
+///
+/// # Safety
+///
+/// This function is unsafe to implement. In particular, the following
+/// must hold:
+/// - The value of either method does not change when either function of this
+///   trait is called, unless `self` is moved.
+/// - `slice::from_raw_parts_mut(as_ptr(), len())` is a valid reference to
+///   a slice.
+/// - Calling either function does not create a reference to is pointee, to
+///   avoid violating the aliasing rules.
+pub unsafe trait Buf {
+    /// Returns a pointer to the buffer of bytes managed by this `Buf`.
+    ///
+    /// This function may be called while there exist outstanding references
+    /// into this buffer.
+    #[allow(clippy::wrong_self_convention)]
+    fn as_ptr(this: &UnsafeCell<Self>) -> *mut u8;
+
+    /// Returns the length of the buffer, in bytes.
+    fn len(this: &UnsafeCell<Self>) -> usize;
+}
+
+unsafe impl Buf for [u8] {
+    #[inline]
+    fn as_ptr(this: &UnsafeCell<Self>) -> *mut u8 {
+        // SAFETY: Casting a fat pointer to a thin pointer always pulls out
+        // the "pointer" part.
+        this.get() as *mut u8
+    }
+
+    #[inline]
+    fn len(this: &UnsafeCell<Self>) -> usize {
+        unsafe {
+            // TODO: This currently materializes a reference to the slice.
+            // We would like to use `<*mut [u8]>::len()`, but that requires
+            // nightly. =(
+            (&*this.get()).len()
+        }
+    }
+}
+
+unsafe impl<const N: usize> Buf for [u8; N] {
+    #[inline]
+    fn as_ptr(this: &UnsafeCell<Self>) -> *mut u8 {
+        // SAFETY: *mut [T; N] as *mut T is a pointer to the first element.
+        this.get() as *mut u8
+    }
+
+    #[inline]
+    fn len(_: &UnsafeCell<Self>) -> usize {
+        N
+    }
+}
+
+// SAFETY: We can materialize the &mut Vec without any trouble, since the
+// aliased memory is the data it points to, which is *not* materialized
+// unless Deref is called, which neither of the functions below do.
+#[cfg(feature = "std")]
+unsafe impl Buf for Vec<u8> {
+    #[inline]
+    fn as_ptr(this: &UnsafeCell<Self>) -> *mut u8 {
+        unsafe { (&mut *this.get()).as_mut_ptr() }
+    }
+
+    #[inline]
+    fn len(this: &UnsafeCell<Self>) -> usize {
+        unsafe { (&*this.get()).len() }
+    }
+}
+
+// SAFETY: Use the fact that Box[T] is representation-equivalent to
+// &mut T (due to FFI safety).
```

Comment:
Is this meant to say `Box<T>` or `Box<[T]>`? In either case, the claim this comment makes is inaccurate:

https://doc.rust-lang.org/std/boxed/index.html#memory-layout

> **So long as T: Sized**, a Box<T> is guaranteed to be represented as a single pointer and is also ABI-compatible with C pointers (i.e. the C type T*)

(emphasis added)

Notably, `[u8]` is not `Sized`, and nor is an arbitrary `T`.

Reply:
Ugh. That's frustrating. I may need to just materialize a `Box` until the relevant `*mut [u8]` methods land in stable...

Reply:
Ugh. I came up with an entirely different way of handling this to avoid all this fuss.